blockchain/addressGenerator_test.go

```diff
@@ -1,69 +1,163 @@
 	"github.com/ElrondNetwork/elrond-sdk-erdgo/data"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-func TestAddress_GetShard(t *testing.T) {
+func TestNewAddressGenerator(t *testing.T) {
 	t.Parallel()
 
-	pubkey := make([]byte, 32)
-	_, _ = rand.Read(pubkey)
-
-	numShardsWithoutMeta := uint32(2)
-	shardCoordinator, _ := NewShardCoordinator(numShardsWithoutMeta, 0)
-
-	pubkey[31] &= 0xFE
-	addr0 := data.NewAddressFromBytes(pubkey)
-
-	pubkey[31] |= 0x01
-	addr1 := data.NewAddressFromBytes(pubkey)
-
-	sh0, err := shardCoordinator.ComputeShardId(addr0)
-	assert.Nil(t, err)
-
-	sh1, err := shardCoordinator.ComputeShardId(addr1)
-	assert.Nil(t, err)
-
-	assert.Equal(t, sh0, uint32(0))
-	assert.Equal(t, sh1, uint32(1))
+	t.Run("nil pubkey converter", func(t *testing.T) {
+		t.Parallel()
+
+		args := ArgsAddressGenerator{
+			PubkeyConv: nil,
+		}
+		ag, err := NewAddressGenerator(args)
+
+		assert.Nil(t, ag)
+		assert.Equal(t, process.ErrNilPubkeyConverter, err)
+	})
+	t.Run("nil address generator core", func(t *testing.T) {
+		t.Parallel()
+
+		args := ArgsAddressGenerator{
+			PubkeyConv: core.AddressPublicKeyConverter,
+		}
+		ag, err := NewAddressGenerator(args)
+
+		assert.Nil(t, ag)
+		assert.Equal(t, ErrNilAddressGenerator, err)
+	})
+	t.Run("should work", func(t *testing.T) {
+		args := ArgsAddressGenerator{
+			PubkeyConv: core.AddressPublicKeyConverter,
+			AddressGeneratorCore: &mock.AddressGeneratorStub{
+				NewAddressCalled: func(address []byte, nonce uint64, vmType []byte) ([]byte, error) {
+					return nil, nil
+				},
+			},
+		}
+		ag, err := NewAddressGenerator(args)
+		require.Nil(t, err)
+		require.NotNil(t, ag)
+	})
 }
 
 func TestGenerateSameDNSAddress(t *testing.T) {
 	t.Parallel()
 
-	coord, err := NewShardCoordinator(3, 0)
-	require.Nil(t, err)
-
-	ag, err := NewAddressGenerator(coord)
-	require.Nil(t, err)
-
-	newDNS, err := ag.CompatibleDNSAddressFromUsername("laura.elrond")
-	require.Nil(t, err)
-
-	fmt.Printf("Compatibile DNS address is %s\n", newDNS.AddressAsBech32String())
-	assert.Equal(t, "erd1qqqqqqqqqqqqqpgqvrsdh798pvd4x09x0argyscxc9h7lzfhqz4sttlatg", newDNS.AddressAsBech32String())
+	t.Run("New address errors should error", func(t *testing.T) {
+		t.Parallel()
+
+		expectedErr := fmt.Errorf("expected error")
+		args := ArgsAddressGenerator{
+			PubkeyConv: core.AddressPublicKeyConverter,
+			AddressGeneratorCore: &mock.AddressGeneratorStub{
+				NewAddressCalled: func(address []byte, nonce uint64, vmType []byte) ([]byte, error) {
+					return nil, expectedErr
+				},
+			},
+		}
+		ag, err := NewAddressGenerator(args)
+		require.Nil(t, err)
+
+		newDNS, err := ag.CompatibleDNSAddressFromUsername("laura.elrond")
+		require.Nil(t, newDNS)
+		assert.Equal(t, expectedErr, err)
+	})
+	t.Run("should work", func(t *testing.T) {
+		expectedScAddress, _ := data.NewAddressFromBech32String("erd1qqqqqqqqqqqqqpgqxcy5fma93yhw44xcmt3zwrl0tlhaqmxrdwpsr2vh8p")
```

Comment:
can you use the previously used address?
why has the address changed?
is there a backwards incompatibility here?

Reply:
done